apps/web/src/hooks/usePermit2Requires.ts

```diff
@@ -0,0 +1,47 @@
+import { CurrencyAmount, Token } from '@pancakeswap/swap-sdk-core'
+import { bscTestnetTokens, ethereumTokens } from '@pancakeswap/tokens'
+import { useMemo } from 'react'
+import { Address, isAddressEqual } from 'viem'
+import useAccountActiveChain from './useAccountActiveChain'
+import useCurrentBlockTimestamp from './useCurrentBlockTimestamp'
+import { usePermit2Allowance } from './usePermit2Allowance'
+import { usePermit2Details } from './usePermit2Details'
+
+export const usePermit2Requires = (amount: CurrencyAmount<Token> | undefined, spender?: Address) => {
+  const { account } = useAccountActiveChain()
+  const { allowance, refetch } = usePermit2Allowance(account, amount?.currency)
+  const { amount: permitAmount, expiration } = usePermit2Details(account, amount?.currency, spender)
+  const now = useCurrentBlockTimestamp() ?? 0n
+
+  const requireRevoke = useMemo((): boolean => {
+    const isMainnetUSDT =
+      amount?.currency?.chainId === ethereumTokens.usdt.chainId &&
+      isAddressEqual(amount?.currency.address, ethereumTokens.usdt.address)
+
+    const isBSCTestNetBUSD =
+      amount?.currency?.chainId === bscTestnetTokens.busd.chainId &&
+      isAddressEqual(amount?.currency.address, bscTestnetTokens.busd.address)
+
+    if (!isMainnetUSDT || isBSCTestNetBUSD) return false
+
+    return !!allowance && allowance.greaterThan(0) && allowance.lessThan(amount)
+  }, [allowance, amount])
+
+  const requireApprove = useMemo((): boolean => {
+    return !!amount && !!allowance && allowance.lessThan(amount)
+  }, [allowance, amount])
+
+  const requirePermit = useMemo((): boolean => {
+    return (amount && permitAmount?.lessThan(amount)) || (Boolean(expiration) && now >= expiration)
```

Comment:
I'd suggest we do this check in real time. When we need to check if the user requires permit, check with the current timestamp. There may be a mismatch between block timestamp and current timestamp, especially on chain with longer block time

Reply:
For instance

```typescript
if (isPermitRequired()) {
  // Then start the permit flow
}
```

Reply:
we may even add a buffer time for that, in case the permit is going to expire in like 30s, and the user probably won't have enough time to do the actual swap

Reply:
I think use the current `requirePermit` is enough, but is a great idea to add a buffer.

because, the steps will only create on user click the swap button, thats say if will check `requireXXX` the same time
just make sure have enough time window for user to execute all actions, so make a enough `buffer` is a good solution